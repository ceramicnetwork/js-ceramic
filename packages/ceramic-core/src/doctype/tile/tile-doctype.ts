import CID from 'cids'

import jsonpatch from 'fast-json-patch'

import { encode as base64Encode } from '@ethersproject/base64'
import { randomBytes } from '@ethersproject/random'

import { DocState, Doctype, InitOpts } from "../../doctype"
import User from "../../user"
import { Context } from "../../context"

const DOCTYPE = 'tile'

/**
 * Tile doctype parameters
 */
export interface TileParams {
    content: object;
    owners?: Array<string>;
}

/**
 * Tile doctype implementation
 */
export class TileDoctype implements Doctype {
    id: string;
    content: object
    doctype: string
    head: CID
    owners: Array<string>
    state: DocState

    /**
     * Create Tile doctype
     * @param params - Create parameters
     * @param context - Ceramic context
     * @param opts - Initialization options
     */
    static async create(params: TileParams, context: Context, opts?: InitOpts): Promise<TileDoctype> {
        if (context.user == null) {
            throw new Error('No user authenticated')
        }

        const { content, owners } = params
        const record = await TileDoctype.makeGenesis(context.user, content, owners, opts)

        const tileDoctype = await context.ceramic.createFromGenesis(record, opts)
        return Promise.resolve(tileDoctype)
    }

    /**
     * Make genesis record
     * @param user - User instance
     * @param content - Context
     * @param owners - Owners
     * @param opts - Initialization parameters
     */
    static async makeGenesis(user: User, content: any, owners?: Array<string>, opts: { isUnique?: boolean } = {}): Promise<any> {
        if (!user) {
            throw new Error('No user authenticated')
        }

        if (!owners) {
            owners = [user.DID]
        }

        let unique: string
        if (opts.isUnique) {
            unique = base64Encode(randomBytes(12))
        }

        const record = { doctype: DOCTYPE, owners, content, unique }
        return TileDoctype._signRecord(record, user)
    }

    /**
     * Change existing Tile doctype
     * @param doctype - Tile doctype instance
     * @param params - Change parameters
     * @param context - Ceramic context
     * @param opts - Initialization options
     */
    static async change(doctype: Doctype, params: TileParams, context: Context, opts?: InitOpts): Promise<Doctype> {
        if (context.user == null) {
            throw new Error('No user authenticated')
        }

        const updateRecord = TileDoctype._makeRecord(doctype, context.user, params.content)
        return await context.ceramic.applyRecord(doctype.id, updateRecord, opts)
    }

    /**
     * Make change record
     * @param doctype - Tile doctype instance
     * @param user - User instance
     * @param newContent - New context
     * @private
     */
    static async _makeRecord(doctype: Doctype, user: User, newContent: any): Promise<Doctype> {
        if (user == null) {
            throw new Error('No user authenticated')
        }

        const patch = jsonpatch.compare(doctype.content, newContent)
        const record: any = { owners: doctype.owners, content: patch, prev: doctype.head, id: doctype.state.log[0] }
        record.iss = user.DID
        // convert CID to string for signing
        const tmpPrev = record.prev
        const tmpId = record.id
        record.prev = { '/': tmpPrev.toString() }
        record.id = { '/': tmpId.toString() }
        const jwt = await user.sign(record, { useMgmt: true})
        const [header, payload, signature] = jwt.split('.') // eslint-disable-line @typescript-eslint/no-unused-vars
        record.prev = tmpPrev
        record.id = tmpId

        return { ...record, header, signature }
    }

    /**
     * Sign Tile record
     * @param user - User instance
     * @param record - Record to be signed
     * @private
     */
    static async _signRecord(record: any, user: User): Promise<any> {
        if (user == null) {
            throw new Error('No user authenticated')
        }
        // TODO - use the dag-jose library for properly encoded signed records
        // The way we use did-jwts right now are quite hacky
        record.iss = user.DID
        // convert CID to string for signing
        const tmpCID = record.prev
        const tmpId = record.id
        if (tmpCID) {
            record.prev = { '/': tmpCID.toString() }
        }
        if (tmpId) {
            record.id = { '/': tmpId.toString() }
        }
        const jwt = await user.sign(record)
        const [header, payload, signature] = jwt.split('.') // eslint-disable-line @typescript-eslint/no-unused-vars
        if (tmpCID) {
            record.prev = tmpCID
        }
        if (tmpId) {
            record.id = tmpId
        }
        return { ...record, header, signature }
    }

}

