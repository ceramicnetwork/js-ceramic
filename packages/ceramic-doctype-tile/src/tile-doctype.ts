import jsonpatch from 'fast-json-patch'

import { encode as base64Encode } from '@ethersproject/base64'
import { randomBytes } from '@ethersproject/random'

import { Doctype, DoctypeConstructor, DoctypeStatic, DocOpts } from "@ceramicnetwork/ceramic-common"
import { Context } from "@ceramicnetwork/ceramic-common"
import { User } from "@ceramicnetwork/ceramic-common"

const DOCTYPE = 'tile'

/**
 * Tile doctype parameters
 */
export interface TileParams {
    content: object;
    owners?: Array<string>;
}

/**
 * Tile doctype implementation
 */
@DoctypeStatic<DoctypeConstructor<TileDoctype>>()
export class TileDoctype extends Doctype {

    /**
     * Change existing Tile doctype
     * @param params - Change parameters
     * @param opts - Initialization options
     */
    async change(params: TileParams, opts: DocOpts = {}): Promise<void> {
        if (this.context.user == null) {
            throw new Error('No user authenticated')
        }

        const updateRecord = await TileDoctype._makeRecord(this, this.context.user, params.content)
        const updated = await this.context.api.applyRecord(this.id, updateRecord, opts)
        this.state = updated.state
    }

    /**
     * Create Tile doctype
     * @param params - Create parameters
     * @param context - Ceramic context
     * @param opts - Initialization options
     */
    static async create(params: TileParams, context: Context, opts?: DocOpts): Promise<TileDoctype> {
        if (context.user == null) {
            throw new Error('No user authenticated')
        }

        const { content, owners } = params
        const record = await TileDoctype.makeGenesis({ content, owners }, context, opts)
        return context.api.createDocumentFromGenesis(record, opts)
    }

    /**
     * Creates genesis record
     * @param params - Create parameters
     * @param context - Ceramic context
     * @param opts - Initialization options
     */
    static async makeGenesis(params: Record<string, any>, context?: Context, opts: DocOpts = {}): Promise<Record<string, any>> {
        if (!context.user) {
            throw new Error('No user authenticated')
        }

        let { owners } = params
        const { content } = params

        if (!owners) {
            owners = [context.user.DID]
        }

        let unique: string
        if (opts.isUnique) {
            unique = base64Encode(randomBytes(12))
        }

        const record = { doctype: DOCTYPE, owners, content, unique }
        return TileDoctype._signRecord(record, context.user)
    }

    /**
     * Make change record
     * @param doctype - Tile doctype instance
     * @param user - User instance
     * @param newContent - New context
     * @private
     */
    static async _makeRecord(doctype: Doctype, user: User, newContent: any): Promise<Doctype> {
        if (!user) {
            throw new Error('No user authenticated')
        }
        const patch = jsonpatch.compare(doctype.content, newContent)
        const record = { owners: doctype.owners, content: patch, prev: doctype.head, id: doctype.state.log[0] }
        return TileDoctype._signRecord(record, user)
    }

    /**
     * Sign Tile record
     * @param user - User instance
     * @param record - Record to be signed
     * @private
     */
    static async _signRecord(record: any, user: User): Promise<any> {
        if (user == null) {
            throw new Error('No user authenticated')
        }
        // TODO - use the dag-jose library for properly encoded signed records
        // The way we use did-jwts right now are quite hacky
        record.iss = user.DID
        // convert CID to string for signing
        const tmpCID = record.prev
        const tmpId = record.id
        if (tmpCID) {
            record.prev = { '/': tmpCID.toString() }
        }
        if (tmpId) {
            record.id = { '/': tmpId.toString() }
        }
        const jwt = await user.sign(record)
        const [header, payload, signature] = jwt.split('.') // eslint-disable-line @typescript-eslint/no-unused-vars
        if (tmpCID) {
            record.prev = tmpCID
        }
        if (tmpId) {
            record.id = tmpId
        }
        return { ...record, header, signature }
    }

}

