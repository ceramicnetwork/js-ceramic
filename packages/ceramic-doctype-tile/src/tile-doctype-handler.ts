import CID from 'cids'

import * as didJwt from 'did-jwt'

import jsonpatch from 'fast-json-patch'

import { TileDoctype, TileParams } from "./tile-doctype"
import {
    AnchorProof, AnchorRecord, AnchorStatus, DocState, DoctypeConstructor, DoctypeHandler, DocOpts, SignatureStatus
} from "@ceramicnetwork/ceramic-common"
import { Context } from "@ceramicnetwork/ceramic-common"

const DOCTYPE = 'tile'

/**
 * Tile doctype handler implementation
 */
export class TileDoctypeHandler implements DoctypeHandler<TileDoctype> {
    /**
     * Gets doctype name
     */
    get name(): string {
        return DOCTYPE
    }

    /**
     * Gets doctype class
     */
    get doctype(): DoctypeConstructor<TileDoctype> {
        return TileDoctype
    }

    /**
     * Create new Tile doctype instance
     * @param params - Create parameters
     * @param context - Ceramic context
     * @param opts - Initialization options
     */
    async create(params: TileParams, context: Context, opts?: DocOpts): Promise<TileDoctype> {
        return TileDoctype.create(params, context, opts)
    }

    /**
     * Applies record (genesis|signed|anchor)
     * @param record - Record
     * @param cid - Record CID
     * @param context - Ceramic context
     * @param state - Document state
     */
    async applyRecord(record: any, cid: CID, context: Context, state?: DocState): Promise<DocState> {
        if (state == null) {
            // apply genesis
            return this._applyGenesis(record, cid, context)
        }

        if (record.proof) {
            const proofRecord = (await context.ipfs.dag.get(record.proof)).value;
            return this._applyAnchor(record, proofRecord, cid, state);
        }

        return this._applySigned(record, cid, state, context)
    }

    /**
     * Applies genesis record
     * @param record - Genesis record
     * @param cid - Genesis record CID
     * @param context - Ceramic context
     * @private
     */
    async _applyGenesis(record: any, cid: CID, context: Context): Promise<DocState> {
        await this._verifyRecordSignature(record, context)
        // TODO - verify genesis record
        return {
            doctype: DOCTYPE,
            owners: record.owners,
            content: record.content,
            nextContent: null,
            signature: SignatureStatus.SIGNED,
            anchorStatus: AnchorStatus.NOT_REQUESTED,
            log: [cid]
        }
    }

    /**
     * Applies signed record
     * @param record - Signed record
     * @param cid - Signed record CID
     * @param state - Document state
     * @param context - Ceramic context
     * @private
     */
    async _applySigned(record: any, cid: CID, state: DocState, context: Context): Promise<DocState> {
        if (!record.id.equals(state.log[0])) {
            throw new Error(`Invalid docId ${record.id}, expected ${state.log[0]}`)
        }
        await this._verifyRecordSignature(record, context)
        state.log.push(cid)
        return {
            ...state,
            signature: SignatureStatus.SIGNED,
            anchorStatus: AnchorStatus.NOT_REQUESTED,
            nextContent: jsonpatch.applyPatch(state.content, record.content).newDocument
        }
    }

    /**
     * Applies anchor record
     * @param record - Anchor record
     * @param proof - Anchor record proof
     * @param cid - Anchor record CID
     * @param state - Document state
     * @private
     */
    async _applyAnchor(record: AnchorRecord, proof: AnchorProof, cid: CID, state: DocState): Promise<DocState> {
        state.log.push(cid)
        let content = state.content
        if (state.nextContent) {
            content = state.nextContent
            delete state.nextContent
        }
        return {
            ...state, content, anchorStatus: AnchorStatus.ANCHORED, anchorProof: proof,
        }
    }

    /**
     * Verifies record signature
     * @param record - Record to be verified
     * @param context - Ceramic context
     * @private
     */
    async _verifyRecordSignature(record: any, context: Context): Promise<void> {
        // reconstruct jwt
        const { header, signature } = record
        delete record.header
        delete record.signature
        let payload = Buffer.from(JSON.stringify({
            doctype: record.doctype,
            owners: record.owners,
            content: record.content,
            unique: record.unique || undefined,
            prev: record.prev ? { '/': record.prev.toString() } : undefined,
            id: record.id ? { '/': record.id.toString() } : undefined,
            iss: record.iss
        })).toString('base64')
        payload = payload.replace(/=/g, '')
        const jwt = [header, payload, signature].join('.')
        try {
            await this.verifyJWT(jwt, { resolver: context.resolver })
        } catch (e) {
            throw new Error('Invalid signature for signed record. ' + e)
        }
    }

    /**
     * Verifies JWT token
     * @param jwt - JWT token
     * @param opts - verification options
     */
    async verifyJWT(jwt: string, opts: any): Promise<void> {
        await didJwt.verifyJWT(jwt, opts)
    }

}
