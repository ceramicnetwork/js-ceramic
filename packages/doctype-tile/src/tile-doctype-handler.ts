import CID from 'cids'

import * as didJwt from 'did-jwt'
import base64url from 'base64url'
import { PublicKey } from 'did-resolver'

import jsonpatch from 'fast-json-patch'
import cloneDeep from 'lodash.clonedeep'

import { TileDoctype, TileParams } from "./tile-doctype"
import {
    AnchorProof,
    AnchorRecord,
    AnchorStatus,
    Context,
    DocOpts,
    DocState,
    RecordType,
    DoctypeConstructor,
    DoctypeHandler,
    DoctypeUtils,
    SignatureStatus
} from "@ceramicnetwork/common"

const DOCTYPE = 'tile'

/**
 * Tile doctype handler implementation
 */
export class TileDoctypeHandler implements DoctypeHandler<TileDoctype> {
    /**
     * Gets doctype name
     */
    get name(): string {
        return DOCTYPE
    }

    /**
     * Gets doctype class
     */
    get doctype(): DoctypeConstructor<TileDoctype> {
        return TileDoctype
    }

    /**
     * Create new Tile doctype instance
     * @param params - Create parameters
     * @param context - Ceramic context
     * @param opts - Initialization options
     */
    async create(params: TileParams, context: Context, opts?: DocOpts): Promise<TileDoctype> {
        return TileDoctype.create(params, context, opts)
    }

    /**
     * Applies record (genesis|signed|anchor)
     * @param record - Record
     * @param cid - Record CID
     * @param context - Ceramic context
     * @param state - Document state
     */
    async applyRecord(record: any, cid: CID, context: Context, state?: DocState): Promise<DocState> {
        if (state == null) {
            // apply genesis
            return this._applyGenesis(record, cid, context)
        }

        if (record.proof) {
            const proofRecord = (await context.ipfs.dag.get(record.proof)).value;
            return this._applyAnchor(record, proofRecord, cid, state);
        }

        return this._applySigned(record, cid, state, context)
    }

    /**
     * Applies genesis record
     * @param record - Genesis record
     * @param cid - Genesis record CID
     * @param context - Ceramic context
     * @private
     */
    async _applyGenesis(record: any, cid: CID, context: Context): Promise<DocState> {
        let payload = record
        const isSigned = DoctypeUtils.isSignedRecord(record)
        if (isSigned) {
            payload = (await context.ipfs.dag.get(record.link)).value
            await this._verifySignature(record, context, payload.header.controllers[0])
        } else if (payload.data !== null) {
            throw Error('Genesis record with contents should always be signed')
        }
        return {
            doctype: DOCTYPE,
            content: payload.data,
            metadata: payload.header,
            signature: isSigned? SignatureStatus.SIGNED : SignatureStatus.GENESIS,
            anchorStatus: AnchorStatus.NOT_REQUESTED,
            log: [{ cid, type: RecordType.GENESIS }]
        }
    }

    /**
     * Applies signed record
     * @param record - Signed record
     * @param cid - Signed record CID
     * @param state - Document state
     * @param context - Ceramic context
     * @private
     */
    async _applySigned(record: any, cid: CID, state: DocState, context: Context): Promise<DocState> {
        await this._verifySignature(record, context, state.metadata.controllers[0])

        const payload = (await context.ipfs.dag.get(record.link)).value
        if (!payload.id.equals(state.log[0].cid)) {
            throw new Error(`Invalid docId ${payload.id}, expected ${state.log[0].cid}`)
        }

        const nextState = cloneDeep(state)

        nextState.signature = SignatureStatus.SIGNED
        nextState.anchorStatus = AnchorStatus.NOT_REQUESTED

        const nonce = payload.header.nonce
        const squash = nonce > 0 && state.next
        if (squash) {
            nextState.log[nextState.log.length-1] = { cid, type: RecordType.SIGNED }
            nextState.next = {
                content: jsonpatch.applyPatch(state.next.content, payload.data).newDocument
            }
        } else {
            nextState.log.push({ cid, type: RecordType.SIGNED })
            nextState.next = {
                content: jsonpatch.applyPatch(state.content, payload.data).newDocument
            }
        }

        if (payload.header) {
            nextState.next.metadata = {...nextState.metadata, ...payload.header}
        }

        if (nonce) {
            nextState.next.metadata = { ...nextState.next.metadata, nonce: nonce }
        }
        return nextState
    }

    /**
     * Applies anchor record
     * @param record - Anchor record
     * @param proof - Anchor record proof
     * @param cid - Anchor record CID
     * @param state - Document state
     * @private
     */
    async _applyAnchor(record: AnchorRecord, proof: AnchorProof, cid: CID, state: DocState): Promise<DocState> {
        state.log.push({ cid, type: RecordType.ANCHOR })
        let content = state.content
        let metadata = state.metadata

        if (state.next?.content) {
            content = state.next.content
            delete state.next.content
        }

        if (state.next?.metadata) {
            metadata = state.next.metadata
            delete state.next.metadata
        }

        delete state.next

        return {
            ...state, content, metadata, anchorStatus: AnchorStatus.ANCHORED, anchorProof: proof,
        }
    }

    /**
     * Verifies record signature
     * @param record - Record to be verified
     * @param context - Ceramic context
     * @param did - DID value
     * @private
     */
    async _verifySignature(record: any, context: Context, did: string): Promise<void> {
        const { payload, signatures } = record
        const { signature,  protected: _protected } = signatures[0]

        const decodedHeader = JSON.parse(base64url.decode(_protected))
        const { kid } = decodedHeader
        if (!kid.startsWith(did)) {
            throw new Error(`Signature was made with wrong DID. Expected: ${did}, got: ${kid}`)
        }

        const { publicKey } = await context.resolver.resolve(kid)
        const jws = [_protected, payload, signature].join('.')
        try {
            await this.verifyJWS(jws, publicKey)
        } catch (e) {
            throw new Error('Invalid signature for signed record. ' + e)
        }
    }

    /**
     * Verifies JWS token
     * @param jws - JWS token
     * @param pubkeys - public key(s)
     */
    async verifyJWS(jws: string, pubkeys: PublicKey[]): Promise<void> {
        await didJwt.verifyJWS(jws, pubkeys)
    }

}
