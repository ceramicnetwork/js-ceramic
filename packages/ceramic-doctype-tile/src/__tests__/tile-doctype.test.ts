import { TileDoctypeHandler } from '../tile-doctype-handler'
import CID from 'cids'

import { User } from "@ceramicnetwork/ceramic-common"

class MockUser implements User {
  DID: string;
  publicKeys: any;
  async auth(): Promise<void> {
    return null
  }
  async sign(): Promise<string> {
    return null
  }
}

import { TileDoctype } from "../tile-doctype"
import { Context } from "@ceramicnetwork/ceramic-common"
jest.mock('did-jwt', () => ({
  // TODO - We should test for when this function throws as well
  verifyJWT: (): any => 'verified'
}))

const cloneDeep = require('lodash.clonedeep') // eslint-disable-line @typescript-eslint/no-var-requires
const { sha256 } = require('js-sha256') // eslint-disable-line @typescript-eslint/no-var-requires
const hash = (data: string): CID => new CID(1, 'sha2-256', Buffer.from('1220' + sha256(data), 'hex'))

const FAKE_CID_1 = new CID('bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_2 = new CID('bafybeig6xv5nwphfmvcnektpnojts44jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_3 = new CID('bafybeig6xv5nwphfmvcnektpnojts55jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_4 = new CID('bafybeig6xv5nwphfmvcnektpnojts66jqcuam7bmye2pb54adnrtccjlsu')

const RECORDS = {
  genesis: { doctype: 'tile', owners: [ 'did:3:bafyasdfasdf' ], content: { much: 'data' } },
  genesisGenerated: { doctype: 'tile', owners: [ 'did:3:bafyasdfasdf' ], content: { much: 'data' }, iss: 'did:3:bafyuser', header: 'aaaa', signature: 'cccc' },
  r1: {
    desiredContent: { much: 'data', very: 'content' },
    record: { content: [ { op: 'add', path: '/very', value: 'content' } ], owners: ["did:3:bafyasdfasdf"], id: FAKE_CID_1, prev: FAKE_CID_1, iss: 'did:3:bafyuser', header: 'aaaa', signature: 'cccc' }
  },
  r2: { record: { proof: FAKE_CID_4 } },
  proof: {
    value: {
      blockNumber: 123456
    }
  }
}

const wrapWithFakeSignature = (obj: any): any => {
  obj.header = 'aaaa'
  obj.signature = 'cccc'
  obj.iss = 'did:3:bafyuser'
  return obj
}

describe('ThreeIdHandler', () => {
  let user: MockUser;
  let tileDoctypeHandler: TileDoctypeHandler;
  let context: Context;

  beforeAll(() => {
    user = new MockUser()
    user.sign = jest.fn(async () => {
      // fake jwt
      return 'aaaa.bbbb.cccc'
    })
    user.DID = 'did:3:bafyuser'

    const recs: Record<string, any> = {}
    const ipfs = {
      dag: {
        put(rec: any, cid?: CID): any {
          if (cid) {
            recs[cid.toString()] = rec
            return cid
          }
          // stringify as a way of doing deep copy
          const clone = cloneDeep(rec)
          const c = hash(JSON.stringify(clone))
          recs[c.toString()] = clone
          return c
        },
        get(cid: any): any {
          return recs[cid.toString()]
        }
      }
    }

    context = {
      ipfs: ipfs,
      anchorService: null,
    }
  })

  beforeEach(() => {
    tileDoctypeHandler = new TileDoctypeHandler()
  })

  it('is constructed correctly', async () => {
    expect(tileDoctypeHandler.name).toEqual('tile')
  })

  it('makes genesis record correctly', async () => {
    const record1 = await TileDoctype.makeGenesis({ content: RECORDS.genesis.content, owners: [user.DID] }, { user })

    const expected1 = wrapWithFakeSignature({ doctype: RECORDS.genesis.doctype, content: RECORDS.genesis.content, owners: [user.DID] })
    expect(record1).toEqual(expected1)

    const record2 = await TileDoctype.makeGenesis({ content: RECORDS.genesis.content, owners: RECORDS.genesis.owners }, { user })
    expect(record2).toEqual(wrapWithFakeSignature(RECORDS.genesis))
  })

  it('creates genesis records deterministically by default', async () => {
    const record1 = await TileDoctype.makeGenesis({ content: RECORDS.genesis.content, owners: [user.DID] }, { user })
    const record2 = await TileDoctype.makeGenesis({ content: RECORDS.genesis.content, owners: [user.DID] }, { user })

    expect(record1).toEqual(record2)
  })

  it('creates a unique genesis record if specified', async () => {
    const record1 = await TileDoctype.makeGenesis({ content: RECORDS.genesis.content, owners: [user.DID] }, { user }, { isUnique: true })
    const record2 = await TileDoctype.makeGenesis({ content: RECORDS.genesis.content, owners: [user.DID] }, { user }, { isUnique: true })

    expect(record1).not.toEqual(record2)
  })

  it('applies genesis record correctly', async () => {
    const tileHandler = new TileDoctypeHandler()

    await context.ipfs.dag.put(RECORDS.genesisGenerated, FAKE_CID_1)

    const docState = await tileHandler.applyRecord(RECORDS.genesis, FAKE_CID_1, context)
    expect(docState).toMatchSnapshot()
  })

  it('makes signed record correctly', async () => {
    const tileDoctypeHandler = new TileDoctypeHandler()

    await context.ipfs.dag.put(RECORDS.genesisGenerated, FAKE_CID_1)

    const state = await tileDoctypeHandler.applyRecord(RECORDS.genesisGenerated, FAKE_CID_1, context)
    const doctype = new TileDoctype(state, context)

    await expect(TileDoctype._makeRecord(doctype, null, RECORDS.r1.desiredContent)).rejects.toThrow(/No user/)

    const record = await TileDoctype._makeRecord(doctype, user, RECORDS.r1.desiredContent)
    expect(record).toEqual(RECORDS.r1.record)
  })

  it('applies signed record correctly', async () => {
    const tileDoctypeHandler = new TileDoctypeHandler()

    await context.ipfs.dag.put(RECORDS.genesisGenerated, FAKE_CID_1)
    await context.ipfs.dag.put(RECORDS.r1.record, FAKE_CID_2)

    let state = await tileDoctypeHandler.applyRecord(RECORDS.genesisGenerated, FAKE_CID_1, context)
    state = await tileDoctypeHandler.applyRecord(RECORDS.r1.record, FAKE_CID_2, context, state)
    expect(state).toMatchSnapshot()
  })

  it('applies anchor record correctly', async () => {
    const tileDoctypeHandler = new TileDoctypeHandler()

    await context.ipfs.dag.put(RECORDS.genesisGenerated, FAKE_CID_1)
    await context.ipfs.dag.put(RECORDS.r1.record, FAKE_CID_2)
    await context.ipfs.dag.put(RECORDS.r2.record, FAKE_CID_3)
    await context.ipfs.dag.put(RECORDS.proof, FAKE_CID_4)

    let state = await tileDoctypeHandler.applyRecord(RECORDS.genesis, FAKE_CID_1, context)
    state = await tileDoctypeHandler.applyRecord(RECORDS.r1.record, FAKE_CID_2, context, state)
    state = await tileDoctypeHandler.applyRecord(RECORDS.r2.record, FAKE_CID_3, context, state)
    expect(state).toMatchSnapshot()
  })
})
