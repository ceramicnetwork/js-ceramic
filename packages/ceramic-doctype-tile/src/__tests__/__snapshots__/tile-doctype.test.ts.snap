// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TileDoctypeHandler applies anchor record correctly 1`] = `
Object {
  "anchorProof": Object {
    "blockNumber": 123456,
  },
  "anchorStatus": 3,
  "content": Object {
    "much": "data",
    "very": "content",
  },
  "doctype": "tile",
  "log": Array [
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          115,
          137,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          119,
          169,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
  ],
  "owners": Array [
    "did:3:bafyasdfasdf",
  ],
  "signature": 2,
}
`;

exports[`TileDoctypeHandler applies genesis record correctly 1`] = `
Object {
  "anchorStatus": 0,
  "content": Object {
    "much": "data",
  },
  "doctype": "tile",
  "log": Array [
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
  ],
  "nextContent": null,
  "owners": Array [
    "did:3:bafyasdfasdf",
  ],
  "signature": 2,
}
`;

exports[`TileDoctypeHandler applies signed record correctly 1`] = `
Object {
  "anchorStatus": 0,
  "content": Object {
    "much": "data",
    "very": "content",
  },
  "doctype": "tile",
  "log": Array [
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          115,
          137,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
  ],
  "nextContent": Object {
    "much": "data",
    "very": "content",
  },
  "owners": Array [
    "did:3:bafyasdfasdf",
  ],
  "signature": 2,
}
`;

exports[`TileDoctypeHandler applies anchor record correctly 1`] = `
Object {
  "anchorProof": Object {
    "blockNumber": 123456,
  },
  "anchorStatus": 3,
  "content": Object {
    "much": "data",
    "very": "content",
  },
  "doctype": "tile",
  "log": Array [
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          115,
          137,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          119,
          169,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
  ],
  "owners": Array [
    "did:3:bafyasdfasdf",
  ],
  "signature": 2,
}
`;

exports[`TileDoctypeHandler applies genesis record correctly 1`] = `
Object {
  "anchorStatus": 0,
  "content": Object {
    "much": "data",
  },
  "doctype": "tile",
  "log": Array [
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
  ],
  "nextContent": null,
  "owners": Array [
    "did:3:bafyasdfasdf",
  ],
  "signature": 2,
}
`;

exports[`TileDoctypeHandler applies signed record correctly 1`] = `
Object {
  "anchorStatus": 0,
  "content": Object {
    "much": "data",
    "very": "content",
  },
  "doctype": "tile",
  "log": Array [
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          115,
          137,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
  ],
  "nextContent": Object {
    "much": "data",
    "very": "content",
  },
  "owners": Array [
    "did:3:bafyasdfasdf",
  ],
  "signature": 2,
}
`;
